### ISSUE TITLE
Uri.resolveUri does not work correctly for relative URIs.

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#27447]
```dart
main() {
  {
    Uri base = Uri.parse('a.dart');
    Uri contained = Uri.parse('../b.dart');
    var resolved = base.resolveUri(contained);
    print(resolved);
  }
  {
    Uri base = Uri.parse('a/b.dart');
    Uri contained = Uri.parse('../../c.dart');
    var resolved = base.resolveUri(contained);
    print(resolved);
  }
}
```

I expected that this code would print `../b.dart` and `../c.dart`.
But it prints `b.dart` and `c.dart`.

Moreover, if I step into `_SimpleUri.resolveUri` and invoke `_toNonSimple().resolveUri(reference)`, I get the expected results. So, I guess that the `_SimpleUri`'s implementation is broken.

@stereotype441 @lrhn

**Comment by lrhn** [comment]
We have special casing for resolving a relative URI reference against another relative URI reference (the URI spec assumes always resolving against a non-relative URI, and it's behavior isn't really useful, but technically we are not following the specification).
It seems the simple-URI implementation fails the special casing here for some reason.

**Comment by mit-mit** [comment]
@scheglov is this really P1?

**Comment by scheglov** [comment]
This causes invalid analysis (and possibly DDC compilation) when using summaries.
This seems pretty serious to me.
But I'm not an expert in priorities.

**Comment by mit-mit** [comment]
In what real-world context are we using a relative base?

**Comment by stereotype441** [comment]
Analyzer summary logic uses a relative base so that it can create portable summaries.  For example, assuming user Joe has the following file structure:

    /home/joe/proj/foo/bar/baz.dart
      import '../qux.dart';
      class C extends B {}
    /home/joe/proj/foo/qux.dart
      export '../thud.dart';
    /home/joe/proj/thud.dart
      class B {}

In the summary for `/home/joe/proj/foo/bar/baz.dart` we want to record that class C extends class B, and that class B came originally from the file `/home/joe/proj/thud.dart`.  But we don't want to encode the absolute path `/home/joe/proj/thud.dart` in the summary file, because that would make the summary unusable if Joe decided to share this code with some other user.  So instead we use the relative path from `/home/joe/proj/foo/bar/baz.dart` to `/home/joe/proj/thud.dart`, which is `../../thud.dart`.  The easiest way to compute this path is to use Uri.resolveUri to resolve `../thud.dart` relative to `../qux.dart`.

**Comment by lrhn** [comment]
It should be fixed on by https://github.com/dart-lang/sdk/commit/9bd4406d4ba9db9cda26089136d5cc16415125b0 now.

**Comment by lrhn** [comment]
Resolving a URI reference against another *URI reference* (and not a URI) is not specified by RFC 3986, but since we use the `Uri` class to represent both, we have to handle it.
We allow it, and handle it reasonably, because it's useful, e.g., for cases like this, so it should work consistently - but the only exception from the RFC 9386 algorithm is in the case where the base URI has neither scheme nor authority and a non-empty relative path.

**Comment by mit-mit** [comment]
The last full push to 1.20 happened earlier today; do we need to merge this? 

CC @whesse

**Comment by stereotype441** [comment]
Yes, that would be very helpful, since analyzer's workaround for this issue is incomplete, and we can't remove it safely until a version of the Dart SDK is published that contains the bug fix.  So if we don't get this fix merged into 1.20, we'll have to wait until 1.21 is released before we can clean up analyzer.

**Comment by scheglov** [comment]
For reference, [this CL](https://codereview.chromium.org/2383643002/) would remove the incomplete analyze workaround.

**Comment by whesse** [comment]
The fix for this was the last thing that made it into -dev.10.0, so no, we don't need to merge this.

**Comment by mit-mit** [comment]
Great, thanks for checking!
